File: php/factory_method.php
<?php

abstract class Product {
	abstract public function show(): string;
}

class ProductOne extends Product {
	public function show(): string {
		return "product one";
	}
}

class ProductTwo extends Product {
	public function show(): string {
		return "product two";
	}
}

abstract class Creator {
	abstract public function factoryMethod(): Product;

	public function showOutput(): void {
		$product = $this->factoryMethod();

		echo "output: {$product->show()}";
	}
}

class ConcreteCreatorProductOne extends Creator {
	public function factoryMethod(): Product {
		return new ProductOne();
	}
}

class ConcreteCreatorProductTwo extends Creator {
	public function factoryMethod(): Product {
		return new ProductTwo();
	}
}

function clientCode(Creator $creator): void {
	$creator->showOutput();
}

clientCode(new ConcreteCreatorProductTwo());

File: criational/factory_method.py
#!/usr/bin/env python3

from abc import ABC, abstractmethod


class Product(ABC):
    @abstractmethod
    def show(self) -> str:
        pass


class ConcreteProductOne(Product):
    def show(self) -> str:
        return "product one"


class ConcreteProductTwo(Product):
    def show(self) -> str:
        return "product two"


class ProductCreator(ABC):
    @abstractmethod
    def factory_method(self):
        pass

    def show_output(self) -> str:
        product = self.factory_method()

        print(f"output: {product.show()}")


class ConcreteProductCreatorOne(ProductCreator):
    def factory_method(self) -> Product:
        return ConcreteProductOne()


class ConcreteProductCreatorTwo(ProductCreator):
    def factory_method(self) -> Product:
        return ConcreteProductTwo()


def execute(creator: ProductCreator) -> None:
    creator.show_output()


execute(ConcreteProductCreatorOne())
execute(ConcreteProductCreatorTwo())
